### DISCUSSION TITLE
IVONORMALS3

### DISCUSSION THREAD

**Discussion by Kjasi** [Markemp/Cryengine-Converter#87] [Technical]
So when testing my IVO fix today, I found something interesting; m_ccc_medium_helmet_01.skinm has both a IVONORMALS and a IVONORMALS3 chunks! (I got a "chunk already in map" error)

So I went looking at the IVONORMALS3 a bit closer, and I don't think they're normals at all. I think it's another Vert Color map in a IRGBA Chunk, which would make 2 color maps in a single model. Even though there are colors in the IVOVERTSUVS, I think that they needed a second set of vert colors for their mapping, so they created the new chunk.

Something to look at...

Additionally, we need a Technical discussion category, for topics like this.

**Comment by Markemp** [comment]
Added a Technical discussion category.

Olde timey Cryengine files did support 2 different color structurs (COLORS and COLORS2).  It's likely that this is the Ivo version of it.

I don't know if Collada can support a 2nd Color chunk, but will check it out.

**Reply by Kjasi** [reply]
Sounds good. I just looked, and it looks like Collada isn't exporting colors when using the tmpVertsUVs. I'm working on that code fix... Just checking my tests... Getting odd results VS the raw file data...

**Reply by Markemp** [reply]
I think there are 2 places where you need to identify a new set of attributes for the vertices.  

![image](https://user-images.githubusercontent.com/12057128/127723486-5edcd16a-ddd6-49ac-b6c6-8eb3089fe60a.png)

I'm not sure if colors needs to be added under the `triangles` element too.   I had colors working at one point, but took it out for some reason.  I think it's time to add it again, but I'm not quite sure how the 2nd color structure will work.  I'll merge your commit in as I don't think it will break anything.  But yeah... colors needs some love.

**Reply by Kjasi** [reply]
Weird... My test export already had the input colors...
![image](https://user-images.githubusercontent.com/5256123/127724377-7b313ea2-33a9-4a46-bfa0-4c18a252e86e.png)
I'll check with the rifle later tonight.

**Reply by Markemp** [reply]
Umm... shoot.  I think there are some bugs where the colors in the triangles aren't working right with VertsUVs.  It's not consistent.  I'm digging through right now to get it to work across all Cryengine versions. 

I'm assuming the helmet you linked is an #ivo file?

**Reply by Kjasi** [reply]
Yes. `Data\Objects\Characters\Human\male_v7\armor\ccc\m_ccc_heavy_armor_helmet_01.skin` to be exact. It also has the 2nd Vert Colors chunk.

**Reply by Markemp** [reply]
Can you upload the dae file to this discussion?  I think files can be attached...

**Reply by Markemp** [reply]
I think I got it sorted out.  Had to add some more structure to the source, but it seems to work.  Will be releasing an updated release candidate for 1.3.1 shortly with all these changes.

**Reply by Kjasi** [reply]
Cool! (Do you still need/want the dae file?)

**Reply by Markemp** [reply]
No, I'm good.  Check out the latest v1.3.1 file and let me know if that works for you.  Thanks!

**Reply by Kjasi** [reply]
Results of my test importing IVO Skin files, and applied RGB Vertex Colors to Diffuse Color. (Alpha to Metallic for 2nd image)
![image](https://user-images.githubusercontent.com/5256123/127748666-130e3649-6004-4d96-8bd3-636052d323d1.png)
![image](https://user-images.githubusercontent.com/5256123/127748875-f88d4635-4833-4724-a048-9e079959ad59.png)

CGA conversion seem to have colors default to 0.0 (black). They should default to 1.0. (white). Tested with `3.13.1\Data\Objects\Spaceships\Ships\DRAK\Buccaneer\Exterior\DRAK_Buccaneer.cga`. (I might take that on...)

And it looks like the "models with two "normals"" error is stopping the model from exporting.
```
There was an error rendering R:\StarCitizen\Extracted\3.13.1\Data\Objects\Characters\Human\male_v7\armor\ccc\m_ccc_medium_helmet_01.skin

An item with the same key has already been added. Key: 6

   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at CgfConverter.CryEngineCore.Chunks.ChunkIvoSkin_900.Read(BinaryReader b)
   at CgfConverter.CryEngineCore.Model.ReadChunks(BinaryReader reader)
   at CgfConverter.CryEngineCore.Model.Load(String fileName)
   at CgfConverter.CryEngine.ProcessCryengineFiles()
   at CgfConverterConsole.Program.Main(String[] args)
********************************************************************************```

**Reply by Kjasi** [reply]
> CGA conversion seem to have colors default to 0.0 (black). They should default to 1.0. (white). Tested with `3.13.1\Data\Objects\Spaceships\Ships\DRAK\Buccaneer\Exterior\DRAK_Buccaneer.cga`. (I might take that on...)

On second thought, maybe not... CIG uses those as masks, not colors... It'll probably be fine as is.

**Reply by Markemp** [reply]
You can quickly check the colors of the vertices by setting the mode in Blender to Vertex paint. :)

The color is definitely set to black for some reason.  I've seen that in other game files too (Hunt, Evolve).

**Reply by Kjasi** [reply]
> The color is definitely set to black for some reason. I've seen that in other game files too (Hunt, Evolve).

I think it's because they use them as masks in the materials.

**Reply by Kjasi** [reply]
Which makes sense for having 2 sets (or more) of vert colors: Having up to 8 channels of masks, or 1 for actual colors and 1 for masks.

**Reply by Kjasi** [reply]
Okay, I wasn't finding any vert colors from non-IVO files. Just submitted #93 to correct this.
